feat: Added documentation for UI unit test component query (#1400)
vaadin/docs: articles/tools/testbench/ui-unit-testing.asciidoc

}
----

[[testbench.uiunit.search-components]]
== Searching for Components

[classname]`UIUnitTest` base class is able to get the instantiated view, but child components may not be always accessible directly.
For example components may be stored in fields with private visibility or they may even not be referenced at all in the view class.

> To overcome this limitations, [classname]`UIUnitTest` provides a query functionality, that lets you scan the UI and filter for the components you need to interact with in test methods.

Reviewer: ```suggestion
To overcome this limitation, [classname]`UIUnitTest` provides a Component query functionality, that lets you search the component tree for the components you need to interact with in test methods.
```

Author: done

> Query object can be obtained by calling the [methodname]`select()` method, providing the type of the component you are searching for.

Reviewer: ```suggestion
The `ComponentQuery` object can be obtained by calling the [methodname]`select()` method, providing the type of the component you are searching for.
```

Author: done

}
----

[[testbench.uiunit.search-components]]
== Searching for Components

> [classname]`UIUnitTest` base class is able to get the instantiated view, but child components may not be always accessible directly.

Reviewer 2: ```suggestion
The [classname]`UIUnitTest` base class is able to get the instantiated view, but child components may not be always accessible directly.
```

For example components may be stored in fields with private visibility or they may even not be referenced at all in the view class.
To overcome this limitation, [classname]`UIUnitTest` provides a Component query functionality, that lets you search the component tree for the components you need to interact with in test methods.

The [classname]`ComponentQuery` object can be obtained by calling the [methodname]`$()` method, providing the type of the component you are searching for.

> You can also restrict search scope to the children of current view, using the [methodname]`$view()` method.

Reviewer 2: ```suggestion
You can also restrict search scope to the children of the current view, using the [methodname]`$view()` method.
```

The query object has many filtering utilities that can be used to refine the search.
For example you can filter by component `id`, by a property value, or using custom predicates on potential candidates.
Once the query is ready with all required conditions configured, you can get test wrappers for specific components found by query execution by calling terminal operators.

> Example of terminal operators are [methodname]`first()`, [methodname]`last()`, [methodname]`atIndex()`, [methodname]`all()` and [methodname]`id()`.

Reviewer 2: ```suggestion
Examples of terminal operators are [methodname]`first()`, [methodname]`last()`, [methodname]`atIndex()`, [methodname]`all()` and [methodname]`id()`.
```

.Using Query API to get component wrappers
[source,java]
----
// Get a wrapper for the first TextField in the UI
TextFieldWrap nameField_ = $(TextField.class).first();

// Get a wrapper for the first TextField in the current view
TextFieldWrap nameField_ = $view(TextField.class).first();

// Get a wrapper for the first TextField with the given label
TextFieldWrap nameField_ = $view(TextField.class)
        .withPropertyValue(TextField::getLabel, "First name")
        .first();

// Get a wrapper for the first TextField that satisfies the conditions
Predicate<TextField> fieldHasNotValue = field -> field.getOptionalValue().isEmpty();
Predicate<TextField> fieldIsInvalid = TextField::isInvalid;
TextFieldWrap<TextField, String> textField = $view(TextField.class)
        .withCondition(fieldHasNotValue.or(fieldIsInvalid))
        .first();

----

> Sometimes you may need to query for components nested inside the UI, in a hierarchy composed by many different types of components.

Reviewer 2: ```suggestion
Sometimes you may need to do a query for components nested inside the UI, in a hierarchy composed of many different types of components.
```

To simplify such situations, the query object offers methods to chain a new query starting on a found component, so that complex query can be created in a fluent way.

> The [methodname]`thenOn()` method and its variants, for example [methodname]`thenOnFirst()`, provides you a new query object for the given component type, setting the search scope to the component selected from current query.

Reviewer 2: ```suggestion
The [methodname]`thenOn()` method and its variants, for example [methodname]`thenOnFirst()`, provide you with a new query object for the given component type, setting the search scope to the component selected from the current query.
```
